File: src/workers/methods/ensrf.js
import { periodicDist, gaspariCohn } from '../math';

export function updateEnSRF(state, y, nobs, N, obsIndices, R_diag, localization) {
  const M = state.ensembleSize;
  const ens = state.ensemble;
  const x_mean = state.x_mean;

  for (let ob = 0; ob < nobs; ob++) {
    const obsIdx = obsIndices[ob];
    const y_val = y[ob];
    const r_val = R_diag;

    let h_mean = 0;
    for (let i = 0; i < M; i++) h_mean += ens[i][obsIdx];
    h_mean /= M;

    const h_dev = new Float64Array(M);
    for (let i = 0; i < M; i++) h_dev[i] = ens[i][obsIdx] - h_mean;

    let hph = 0;
    for (let i = 0; i < M; i++) hph += h_dev[i] * h_dev[i];
    hph /= (M - 1);

    const K = new Float64Array(N);
    for (let i = 0; i < N; i++) {
      let cov = 0;
      for (let j = 0; j < M; j++) {
        cov += (ens[j][i] - x_mean[i]) * h_dev[j];
      }
      cov /= (M - 1);
      K[i] = cov * gaspariCohn(periodicDist(i, obsIdx, N), localization) / (hph + r_val);
    }

    const alpha = 1.0 / (1.0 + Math.sqrt(r_val / (hph + r_val)));

    const innov = y_val - h_mean;

    // Update ensemble: avoid allocating dev_j array
    // x_mean_old is needed - save current x_mean
    const x_mean_old = new Float64Array(N);
    for (let i = 0; i < N; i++) x_mean_old[i] = x_mean[i];
    for (let i = 0; i < N; i++) x_mean[i] += K[i] * innov;

    for (let j = 0; j < M; j++) {
      for (let i = 0; i < N; i++) {
        const dev_ji = ens[j][i] - x_mean_old[i];
        ens[j][i] = x_mean[i] + dev_ji - K[i] * alpha * h_dev[j];
      }
    }
  }
}

File: crates/eduda_wasm/src/methods/ensrf.rs
use crate::math::{periodic_dist, gaspari_cohn};

pub fn update_ensrf(
    ensemble: &mut [f64], // M * N flat slice
    x_mean: &mut [f64],
    y: &[f64],
    obs_indices: &[usize],
    r_diag: f64,
    localization: f64,
    m: usize,
    n: usize,
) {
    let nobs = obs_indices.len();
    if nobs == 0 { return; }

    let inv_m = 1.0 / m as f64;
    let inv_m1 = 1.0 / (m - 1) as f64;

    let mut h_dev = vec![0.0; m];
    let mut k_gain = vec![0.0; n];
    let mut x_mean_old = vec![0.0; n];

    for ob in 0..nobs {
        let obs_idx = obs_indices[ob];
        let y_val = y[ob];

        let mut h_mean = 0.0;
        for j in 0..m {
            h_mean += ensemble[j * n + obs_idx];
        }
        h_mean *= inv_m;

        for j in 0..m {
            h_dev[j] = ensemble[j * n + obs_idx] - h_mean;
        }

        let mut hph = 0.0;
        for j in 0..m {
            hph += h_dev[j] * h_dev[j];
        }
        hph *= inv_m1;

        let denom = hph + r_diag;
        let alpha = 1.0 / (1.0 + (r_diag / denom).sqrt());

        for i in 0..n {
            let mut cov = 0.0;
            for j in 0..m {
                cov += (ensemble[j * n + i] - x_mean[i]) * h_dev[j];
            }
            cov *= inv_m1;
            let gloc = gaspari_cohn(periodic_dist(i, obs_idx, n), localization);
            k_gain[i] = cov * gloc / denom;
        }

        let innov = y_val - h_mean;
        x_mean_old.copy_from_slice(x_mean);
        for i in 0..n {
            x_mean[i] += k_gain[i] * innov;
        }

        for j in 0..m {
            for i in 0..n {
                let dev_ji = ensemble[j * n + i] - x_mean_old[i];
                ensemble[j * n + i] = x_mean[i] + dev_ji - k_gain[i] * alpha * h_dev[j];
            }
        }
    }
}
